### PR TITLE
go-task: init at 2.8.1

### PR DESCRIPTION
**Pull request by Parasrah**
###### Motivation for this change

to add `go-task` to nixpkgs (a make alternative). This is my first time contributing to `nixpkgs` so please let me know if there is anything I'm missing or should have done differently. I didn't try building from source because it appears that `go-task` uses itself in the build

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix
Comment (danieldk) [inline maintainers/maintainer-list.nix:6168]: You should add yourself to the maintainers in a separate commit with the message

```
maintainers: add parasrah
```

File: pkgs/development/tools/go-task/default.nix
Comment (danieldk) [inline pkgs/development/tools/go-task/default.nix:14]: Compiling from source is strongly preferred in nixpkgs. Since this looks like a regular Go project, you could use `buildGoModule` (or `buildGoPackage`).

If you build from source, since the project is on GitHub, you can use `fetchFromGitHub` in place of `fetchurl`.

File: pkgs/development/tools/go-task/default.nix
Comment (danieldk) [inline pkgs/development/tools/go-task/default.nix:25]:
```suggestion
  meta = with stdenv.lib; {
```

File: pkgs/development/tools/go-task/default.nix
Comment (danieldk) [inline pkgs/development/tools/go-task/default.nix:28]: With the change above, you can change this to:

```suggestion
    license     = licenses.mit;
```

File: pkgs/development/tools/go-task/default.nix
Comment (danieldk) [inline pkgs/development/tools/go-task/default.nix:30]:
```suggestion
    maintainers = [ maintainers.parasrah ];
```

**Review by danieldk — CHANGES_REQUESTED** [review]
Thanks you for your first contribution :+1: . I have added some comments on how you could improve the derivation.

File: maintainers/maintainer-list.nix
Comment (Parasrah) [reply maintainers/maintainer-list.nix:6168]: done :+1:

File: pkgs/development/tools/go-task/default.nix
Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix]: done :+1:

File: pkgs/development/tools/go-task/default.nix
Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix]: done :+1:

File: pkgs/development/tools/go-task/default.nix
Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix]: done :+1:

File: pkgs/development/tools/go-task/default.nix
Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix:14]: Yeah that makes total sense. I am a bit confused about how to go about this for this particular package, as it leverages [GoReleaser](https://goreleaser.com/#continuous_integration) for its build. I will try to find some time to dig into it and see if I can't get it working with `buildGoPackage` instead. Thanks!

Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix:14]: I tried with `buildGoModule` first, as it does seem to be a go module, but it was unsuccessful in building `subPackages`. I tried to do it with `buildGoPackage` instead (leveraging `vgo2nix`), and it seems like it has mostly worked, but one thing I've noticed is when running `task --version` it produces `Task version: master` instead of the actual version of the tool. From looking into it a little, it seems like `GoReleaser` pulls the version from the most recent tag and injects it somehow into the build. I'm not terribly familiar with golang packages unfortunately and I haven't been able to determine how it does that, or how to replicate that behavior.

Comment (danieldk) [reply pkgs/development/tools/go-task/default.nix:14]: The version is `cmd/task/task.go`: https://github.com/go-task/task/blob/86be13ff1f1173479d1c032bc1b9eae67f06f64e/cmd/task/task.go#L19

You should be able to use `substituteInPlace` to replace that by an actual version in `postPatch`. E.g. (untested):

```
  postPatch = ''
    substituteInPlace cmd/task/task.go \
      'version = "master"' 'version = "${version}"'
  '';
```

Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix]: Thanks! That indeed did do the trick!

File: pkgs/development/tools/go-task/default.nix
Comment (danieldk) [inline pkgs/development/tools/go-task/default.nix:11]: I looked at the upstream project and they include `go.sum` and go.mod`, so `buildGoModule` is probably the better choice here. Allows you to remove `deps.nix` as well.

File: pkgs/development/tools/go-task/default.nix
Comment (danieldk) [inline pkgs/development/tools/go-task/default.nix:35]:
```suggestion
    platforms   = platforms.all;
```

Most Go programs are multiplatform. Any reason to restrict to x86_64 Linux? Try `platforms.all` and see if see builds it on AArch64 and Darwin.

**Review by danieldk — CHANGES_REQUESTED** [review]
Thank you for making all the changes! Now it's almost in a shape to be merged. I added two more suggestions.

Also, the formatting looks a bit unconventional. Not a blocker, but you can format the derivation using standard nixpkgs formatting using the `nixpkgs-fmt` utility.

File: pkgs/development/tools/go-task/default.nix
Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix]: Good catch, I had that when it was originally pulling the linux tarball, fixed

File: pkgs/development/tools/go-task/default.nix
Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix:11]: I did originally try that, it was complaining about failing to build `subPackages`. I assumed that was something to do with the unconventional build tool. I'm not a go dev though, so I'm not sure

**Comment by Parasrah** [comment]
I updated the formatting too, was previously looking at `godot/default.nix` as a reference for the style. I didn't know about `nixpkgs-fmt`, really happy to discover there is a formatter for nix!

File: pkgs/development/tools/go-task/default.nix
Comment (danieldk) [reply pkgs/development/tools/go-task/default.nix]: I have changed the derivation to use `buildGoModule`. It's only a small change compared to what you had before.

**Comment by danieldk** [comment]
I only have one more small worry before merging this. The `go-task` binary is `task`, which is the same name as the binary for `taskwarrior`. This means that users would not be able to install both packages in their profile side by side. Is this a tool people would typically install in their profile or only in e.g. in the `nativeBuildInputs` of some other derivation? If the former, we may consider renaming the binary to e.g. `go-task`.

**Comment by Parasrah** [comment]
that's a good point, I only started using the tool myself a few weeks ago, but I have it installed via `nix-shell` for the projects and use it mostly for development. I could definitely see someone having it installed in their profile side by side task warrior. It would probably be best to rename the binary, I will see if I can't get that sorted

File: pkgs/development/tools/go-task/default.nix
Comment (Parasrah) [reply pkgs/development/tools/go-task/default.nix]: Ahh so it was supposed to be `subPackages = [ "cmd/task" ];`. Sorry I don't have much experience with golang, I probably should have chosen something more familiar for my first contribution

**Comment by danieldk** [comment]
> that's a good point, I only started using the tool myself a few weeks ago, but I have it installed via `nix-shell` for the projects and use it mostly for development. I could definitely see someone having it installed in their profile side by side task warrior. It would probably be best to rename the binary, I will see if I can't get that sorted

Looks great now! I have pushed one more small change (a way of handling the version that is more canonical in `nixpkgs`). I'll merge the PR when ofborg is happy.

**Review by danieldk — APPROVED** [review]
Congrats with your first derivation in nixpkgs :tada: 

Result of `nixpkgs-review pr 93619` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
<br>- go-task
</details>

**Pull request merged**
